src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -171,6 +171,15 @@ def _encode_subject_alt_name(backend, san):
             )
             assert obj != backend._ffi.NULL
             gn.d.registeredID = obj
+        elif isinstance(alt_name, x509.DirectoryName):
+            gn = backend._lib.GENERAL_NAME_new()
+            assert gn != backend._ffi.NULL
+            name = _encode_name(backend, alt_name.value)
+            # _encode_name registers the X509_NAME for gc so we'll duplicate
+            # a new one that is not gc'd for the struct
+            name = backend._lib.X509_NAME_dup(name)
```

Comment:
What about changing _encode_name to take a `gc` boolean arg? No default, so current call sites would become `_encode_name(backend, <data>, gc=True)`

Whatever pattern we choose will be duplicated for `_encode_asn1_str` and maybe `_txt2obj` as well.